File: front-end/src/app/services/account.service.ts
import { HttpClient } from '@angular/common/http';
import { Injectable, inject } from '@angular/core';
import { Observable } from 'rxjs';
import { Lead } from '../models/lead';

@Injectable({
  providedIn: 'root',
})
export class AccountService {
  public http = inject(HttpClient);
  public apiUrl = 'http://localhost:8080/accounts';

  public getAllLeadsFromAccount(accountId: string): Observable<Lead[]> {
    return this.http.get<Lead[]>(`${this.apiUrl}/${accountId}/leads`);
  }

  public getLeadByIdFromAccount(accountId: string, leadId: string): Observable<Lead> {
    return this.http.get<Lead>(`${this.apiUrl}/${accountId}/leads/${leadId}`);
  }

  public addLeadWithAccount(accountId: string, data: any): Observable<Lead> {
    return this.http.post<Lead>(`${this.apiUrl}/${accountId}/leads`, data);
  }

  public updateLeadFromAccount(accountId: string, leadId: string, data: any): Observable<Lead> {
    return this.http.put<Lead>(`${this.apiUrl}/${accountId}/leads/${leadId}`, data);
  }

  public deleteLeadFromAccount(accountId: string, leadId: string): Observable<Lead> {
    return this.http.delete<Lead>(`${this.apiUrl}/${accountId}/leads/${leadId}`);
  }
}

File: back-end/src/main/java/com/example/lead/service/AccountService.java
package com.example.lead.service;

import java.util.List;

import org.springframework.http.ResponseEntity;

import com.example.lead.entity.Account;
import com.example.lead.entity.Lead;

public interface AccountService {
    ResponseEntity<List<Account>> getAll();

    ResponseEntity<Account> getById(Long id);

    ResponseEntity<Account> add(Account user, String userId);

    ResponseEntity<Account> update(Long id, Account user);

    ResponseEntity<Account> delete(Long id);

    ResponseEntity<String> addUserToAccount(Long accountId, Long userId);

    ResponseEntity<String> removeUserFromAccount(Long accountId, Long userId);

    ResponseEntity<Lead> addLeadWithAccount(Long accountId, Lead lead);

    ResponseEntity<List<Lead>> getAllLeadsFromAccount(Long accountId);

    ResponseEntity<Lead> getLeadByIdFromAccount(Long accountId, Long leadId);

    ResponseEntity<Lead> updateLeadFromAccount(Long accountId, Long leadId, Lead lead);

    ResponseEntity<Lead> deleteLeadFromAccount(Long accountId, Long leadId);
}
